spark/agent/src/main/scala/za/co/absa/spline/harvester/AttributeDependencySolver.scala

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.harvester
+
+import java.util.UUID
+
+import za.co.absa.spline.harvester.ModelConstants.OperationParams._
+import za.co.absa.spline.harvester.ModelConstants.OperationExtras._
+import za.co.absa.spline.model.expr.{Alias, AttrRef, Expression, Generic}
+import za.co.absa.spline.producer.model.{DataOperation, OperationLike}
+
+import scala.collection.mutable
+
+
+case class AttributeDependencies(attributes: Map[UUID, Set[UUID]], operations: Map[UUID, Set[Int]])
+
+case class IOSchema(in: Seq[UUID], out: Seq[UUID]) {
+  def both: Seq[UUID] = in.union(out).distinct
+}
+
+object AttributeDependencySolver {
+
+  def resolveDependencies(dataOperations: Seq[DataOperation], ioOperations: Seq[OperationLike]):
+      AttributeDependencies = {
+
+    val allOperations = dataOperations ++ ioOperations
+    val schemaMap = schemaMapOf(allOperations)
+
+    val attributeDependencies = dataOperations
+      .map(resolveDependencies(_, schemaMap))
+      .reduceOption(mergeDependencies)
+      .getOrElse(Map.empty)
+
+    val operationDependencies = resolveAttributeOperationDependencies(allOperations, attributeDependencies, schemaMap)
+
+    AttributeDependencies(attributeDependencies, operationDependencies)
+  }
+
+  private def resolveDependencies(op: DataOperation, schemaMap: Map[Int, IOSchema]): Map[UUID, Set[UUID]] =
+    op.extra(Name) match {
+      case "Project" => resolveExpressionList(op.params(Transformations), schemaMap(op.id).out)
+      case "Aggregate" => resolveExpressionList(op.params(Aggregations), schemaMap(op.id).out)
+      case "SubqueryAlias" => resolveSubqueryAlias(schemaMap(op.id).in, schemaMap(op.id).out)
+      case "Generate" => resolveGenerator(op)
+      case _ => Map.empty
+    }
+
+  private def resolveExpressionList(exprList: Any, schema: Seq[UUID]): Map[UUID, Set[UUID]] = {
+    exprList
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Usage of get on optional type.](https://app.codacy.com/app/ABSA_OSS/spline/pullRequest?prid=4630741)